MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -0,0 +1,2853 @@
+  // describe every nonprotected image, we must have 3 * <number of nonprotected images> additional descriptors.
+  {
+    AdditionalRecordsForImages = ((2 * mImagePropertiesPrivate.CodeSegmentCountMax + 3) * mImagePropertiesPrivate.ImageRecordCount) +
+                                 (mNonProtectedImageRangesPrivate.NonProtectedImageCount * 3);
+  }
+
+  // STEP 6: Determine the Size of the Final Memory Map
+  //
+  // The final memory map buffer size will include:
+  //
+  // 1. The size required to describe the EFI memory map returned from GetMemoryMap()
+  // 2. The size required to populate entries which are not in the EFI memory map but are included
+  //    in the full address space described by the GCD memory map (which inherits its address width
+  //    from the CPU info HOB).
+  // 3. The size required to split all protected and non-protected image ranges into their own descriptors.
+  // 4. The size required to split all special regions into their own descriptors (done by multiplying
+  //    the memory map size by 2 to accomodate the worst-case scenario).
+  {
+    *MemoryMapSize           = (*MemoryMapSize * 2) + ((*DescriptorSize) * AdditionalRecordsForImages);
+    FinalMemoryMapBufferSize = *MemoryMapSize;
+  }
+
+  // STEP 7: Allocate Memory for the Final Memory Map
+  {
+    FreePool (*MemoryMap);
+    *MemoryMap = AllocateZeroPool (FinalMemoryMapBufferSize);
+
+    if (*MemoryMap == NULL) {
+      Status = EFI_OUT_OF_RESOURCES;
+      ASSERT_EFI_ERROR (Status);
+      goto Cleanup;
+    }
+  }
+
+  // STEP 8: Create a Bitmap to Track Which Descriptors Have Been Updated with Attributes
+  //
+  // The pool allocated for the bitmap will be large enough to hold a bit for every descriptor
+  // in the expanded memory map, but in almost all cases the bitmap will be smaller than this.
+  // This bitmap must be freed after the memory map is no longer in use to avoid changing the
+  // real memory map layout.
+  {
+    NumMemoryMapDescriptors = FinalMemoryMapBufferSize / *DescriptorSize;
+    NumBitmapEntries        = (NumMemoryMapDescriptors % 8) == 0 ? NumMemoryMapDescriptors : (((NumMemoryMapDescriptors / 8) * 8) + 8);
+    mBitmapGlobal           = AllocateZeroPool (NumBitmapEntries / 8);
+
+    if (mBitmapGlobal == NULL) {
+      Status = EFI_OUT_OF_RESOURCES;
+      ASSERT_EFI_ERROR (Status);
+      goto Cleanup;
+    }
+  }
+
+  // STEP 9: Get the EFI Memory Map
+  //
+  // From this point onward there will be no more allocations or frees within this function
+  // to ensure the memory map is not changed during the creation process. This block fetches
+  // the current EFI memory map now that it has stabilized.
+  {
+    Status = CoreGetMemoryMap (
+               MemoryMapSize,
+               *MemoryMap,
+               &MapKey,
+               DescriptorSize,
+               &DescriptorVersion
+               );
+
+    ASSERT_EFI_ERROR (Status);

```

Comment:
Does it make sense to continue if getting the EFI memory map fails here? It doesn't seem like it.

Reply:
Agreed, will fail gracefully if so.